File: frontend/src/api/display.ts
import { apiGet, apiPost, apiPostEmpty, apiPut } from './http'
import {
  asArray,
  asBackgroundFillMode,
  asBoolean,
  asDisplayTransitionMode,
  asFitMode,
  asNumber,
  asOptionalString,
  asRecord,
  asString,
  asStringArray,
  type BackgroundFillMode,
  type DisplayConfig,
  type DisplayConfigUpdateRequest,
  type DisplayPlaylist,
  type PlaybackProgressResponse,
  type PlaylistItem,
  type PlaylistItemBackground,
} from './types'
import { normalizeSleepSchedule } from './settings'

const DEFAULT_DISPLAY_CONFIG: DisplayConfig = {
  id: 'default-display-profile',
  name: 'Default Display',
  selected_collection_id: 'default-collection',
  slideshow_interval_seconds: 30,
  transition_mode: 'slide',
  transition_duration_ms: 700,
  fit_mode: 'contain',
  shuffle_enabled: true,
  shuffle_bag_enabled: true,
  idle_message: 'Add photos from the admin UI to begin playback.',
  refresh_interval_seconds: 60,
  background_fill_mode: 'black',
  is_default: true,
  created_at: '',
  updated_at: '',
}

export async function getDisplayConfig(): Promise<DisplayConfig> {
  const payload = await apiGet<unknown>('/api/display/config')
  return normalizeDisplayConfig(payload)
}

export async function updateDisplayConfig(request: DisplayConfigUpdateRequest): Promise<DisplayConfig> {
  const payload = await apiPut<DisplayConfigUpdateRequest, unknown>('/api/display/config', request)
  return normalizeDisplayConfig(payload)
}

export async function getDisplayPlaylist(): Promise<DisplayPlaylist> {
  const payload = await apiGet<unknown>('/api/display/playlist')
  const record = asRecord(payload)

  return {
    collection_id: asOptionalString(record?.collection_id) ?? null,
    collection_name: asOptionalString(record?.collection_name) ?? null,
    shuffle_enabled: asBoolean(record?.shuffle_enabled, DEFAULT_DISPLAY_CONFIG.shuffle_enabled),
    playlist_revision: asString(record?.playlist_revision, 'empty'),
    playback_mode: asString(record?.playback_mode, ''),
    playback_cycle_id: asString(record?.playback_cycle_id, ''),
    playback_position: asNumber(record?.playback_position, 0),
    profile: record?.profile ? normalizeDisplayConfig(record.profile) : DEFAULT_DISPLAY_CONFIG,
    items: asArray(record?.items, normalizePlaylistItem).filter((item) => item.display_url.length > 0),
    sleep_schedule: record?.sleep_schedule ? normalizeSleepSchedule(record.sleep_schedule) : null,
  }
}

export async function getNewAssetsCount(since: string, collectionId?: string | null): Promise<{ new_assets_count: number; since: string }> {
  const params = new URLSearchParams({ since })
  if (collectionId) {
    params.set('collection_id', collectionId)
  }
  return apiGet(`/api/display/new-assets/count?${params.toString()}`)
}

/**
 * Advance the server-owned playback cursor past a photograph the display has just shown (ADR 0024).
 *
 * Best effort by design: a failed report can cost at most one repeated photograph on the next
 * resume, so it must never surface an error into slideshow playback.
 */
export async function reportPlaybackProgress(
  assetId: string,
  collectionId?: string | null,
  playbackCycleId?: string | null,
): Promise<boolean> {
  try {
    // playback_cycle_id lets the server ignore reports from a display that missed a cycle
    // rollover, so a stale tab cannot advance the fresh pass (ADR 0024).
    const response = await apiPost<
      { asset_id: string; collection_id?: string | null; playback_cycle_id?: string | null },
      PlaybackProgressResponse
    >('/api/display/playlist/progress', {
      asset_id: assetId,
      ...(collectionId ? { collection_id: collectionId } : {}),
      ...(playbackCycleId ? { playback_cycle_id: playbackCycleId } : {}),
    })
    return response.accepted === true
  } catch {
    return false
  }
}

export async function refreshDisplayCache(): Promise<void> {
  await apiPostEmpty<unknown>('/api/display/refresh')
}

export function getDefaultDisplayConfig(): DisplayConfig {
  return DEFAULT_DISPLAY_CONFIG
}

function normalizeDisplayConfig(payload: unknown): DisplayConfig {
  const record = asRecord(payload)

  return {
    id: asString(record?.id, DEFAULT_DISPLAY_CONFIG.id),
    name: asString(record?.name, DEFAULT_DISPLAY_CONFIG.name),
    selected_collection_id: asOptionalString(record?.selected_collection_id) ?? null,
    slideshow_interval_seconds: asNumber(record?.slideshow_interval_seconds, DEFAULT_DISPLAY_CONFIG.slideshow_interval_seconds),
    transition_mode: asDisplayTransitionMode(record?.transition_mode, DEFAULT_DISPLAY_CONFIG.transition_mode),
    transition_duration_ms: asNumber(record?.transition_duration_ms, DEFAULT_DISPLAY_CONFIG.transition_duration_ms),
    fit_mode: asFitMode(record?.fit_mode, DEFAULT_DISPLAY_CONFIG.fit_mode),
    shuffle_enabled: asBoolean(record?.shuffle_enabled, DEFAULT_DISPLAY_CONFIG.shuffle_enabled),
    shuffle_bag_enabled: asBoolean(record?.shuffle_bag_enabled, DEFAULT_DISPLAY_CONFIG.shuffle_bag_enabled),
    idle_message: asString(record?.idle_message, DEFAULT_DISPLAY_CONFIG.idle_message),
    refresh_interval_seconds: asNumber(record?.refresh_interval_seconds, DEFAULT_DISPLAY_CONFIG.refresh_interval_seconds),
    background_fill_mode: asBackgroundFillMode(record?.background_fill_mode, DEFAULT_DISPLAY_CONFIG.background_fill_mode),
    is_default: asBoolean(record?.is_default, DEFAULT_DISPLAY_CONFIG.is_default),
    created_at: asString(record?.created_at, DEFAULT_DISPLAY_CONFIG.created_at),
    updated_at: asString(record?.updated_at, DEFAULT_DISPLAY_CONFIG.updated_at),
  }
}

function normalizePlaylistItemBackground(value: unknown): PlaylistItemBackground | null {
  const record = asRecord(value)
  if (!record) {
    return null
  }

  return {
    ready: asBoolean(record.ready, false),
    dominant_color: asOptionalString(record.dominant_color) ?? null,
    gradient_colors: Array.isArray(record.gradient_colors) ? asStringArray(record.gradient_colors) : null,
  }
}

function normalizePlaylistItem(item: unknown, index: number): PlaylistItem {
  const record = asRecord(item)

  return {
    asset_id: asString(record?.asset_id ?? record?.id, `${index}`),
    filename: asString(record?.filename, `Photo ${index + 1}`),
    display_url: asString(record?.display_url ?? record?.image_url ?? record?.url, ''),
    thumbnail_url: asString(record?.thumbnail_url, ''),
    width: asNumber(record?.width, 0),
    height: asNumber(record?.height, 0),
    checksum_sha256: asString(record?.checksum_sha256, ''),
    mime_type: asString(record?.mime_type, 'image/jpeg'),
    collection_name: asOptionalString(record?.collection_name),
    source_name: asOptionalString(record?.source_name),
    background: normalizePlaylistItemBackground(record?.background),
  }
}

export function buildBackgroundStyle(
  mode: BackgroundFillMode,
  background: PlaylistItemBackground | null | undefined,
): string {
  if (mode === 'black' || !background?.ready) {
    return '#000'
  }

  if (mode === 'palette_wash') {
    const palette = getBackgroundPalette(background)
    if (palette.length >= 3) {
      return [
        `radial-gradient(circle at 18% 24%, ${withAlpha(palette[0], 0.34)} 0%, transparent 50%)`,
        `radial-gradient(circle at 82% 20%, ${withAlpha(palette[1], 0.3)} 0%, transparent 46%)`,
        `radial-gradient(circle at 50% 78%, ${withAlpha(palette[2], 0.28)} 0%, transparent 52%)`,
        `linear-gradient(145deg, ${withAlpha(palette[0], 0.54)}, ${withAlpha(palette[1], 0.42)}, ${withAlpha(palette[2], 0.48)})`,
      ].join(', ')
    }
  }

  if (
    mode === 'gradient' ||
    mode === 'blurred_backdrop' ||
    mode === 'mirrored_edges' ||
    mode === 'soft_vignette' ||
    mode === 'adaptive_auto'
  ) {
    const colors = background.gradient_colors
    if (colors && colors.length >= 2) {
      return `linear-gradient(135deg, ${colors.join(', ')})`
    }
    return background.dominant_color ?? '#000'
  }

  return background.dominant_color ?? '#000'
}

function getBackgroundPalette(background: PlaylistItemBackground): string[] {
  const colors = background.gradient_colors?.filter((color) => color.length > 0) ?? []
  const dominant = background.dominant_color

  if (colors.length >= 3) {
    return colors.slice(0, 3)
  }

  if (colors.length === 2 && dominant) {
    return [colors[0], colors[1], dominant]
  }

  if (colors.length === 1 && dominant) {
    return [colors[0], dominant, dominant]
  }

  if (dominant) {
    return [dominant, dominant, dominant]
  }

  return colors
}

function withAlpha(color: string, alpha: number): string {
  const normalized = color.trim()
  if (normalized.startsWith('#')) {
    const hex = normalized.slice(1)
    const expanded = hex.length === 3 ? hex.split('').map((value) => value + value).join('') : hex
    if (expanded.length === 6) {
      const red = Number.parseInt(expanded.slice(0, 2), 16)
      const green = Number.parseInt(expanded.slice(2, 4), 16)
      const blue = Number.parseInt(expanded.slice(4, 6), 16)
      if ([red, green, blue].every(Number.isFinite)) {
        return `rgba(${red}, ${green}, ${blue}, ${alpha})`
      }
    }
  }

  return normalized
}

File: backend/app/schemas/display.py
from __future__ import annotations

from pydantic import BaseModel, Field, field_validator

from app.models.display import DisplayPlaylist, DisplayProfile, PlaylistItem
from app.schemas.settings import SleepScheduleResponse
from app.services.background_service import VALID_BACKGROUND_FILL_MODES


class DisplayProfileResponse(BaseModel):
    id: str
    name: str
    selected_collection_id: str | None
    slideshow_interval_seconds: int
    transition_mode: str
    transition_duration_ms: int
    fit_mode: str
    shuffle_enabled: bool
    shuffle_bag_enabled: bool
    idle_message: str
    refresh_interval_seconds: int
    is_default: bool
    created_at: str
    updated_at: str
    background_fill_mode: str

    @classmethod
    def from_domain(cls, profile: DisplayProfile) -> "DisplayProfileResponse":
        return cls(
            id=profile.id,
            name=profile.name,
            selected_collection_id=profile.selected_collection_id,
            slideshow_interval_seconds=profile.slideshow_interval_seconds,
            transition_mode=profile.transition_mode,
            transition_duration_ms=profile.transition_duration_ms,
            fit_mode=profile.fit_mode,
            shuffle_enabled=profile.shuffle_enabled,
            shuffle_bag_enabled=profile.shuffle_bag_enabled,
            idle_message=profile.idle_message,
            refresh_interval_seconds=profile.refresh_interval_seconds,
            is_default=profile.is_default,
            created_at=profile.created_at,
            updated_at=profile.updated_at,
            background_fill_mode=profile.background_fill_mode,
        )


class DisplayConfigUpdateRequest(BaseModel):
    name: str | None = None
    selected_collection_id: str | None = None
    slideshow_interval_seconds: int | None = Field(default=None, ge=1, le=3600)
    transition_mode: str | None = None
    transition_duration_ms: int | None = Field(default=None, ge=0, le=30000)
    fit_mode: str | None = None
    shuffle_enabled: bool | None = None
    shuffle_bag_enabled: bool | None = None
    idle_message: str | None = None
    refresh_interval_seconds: int | None = Field(default=None, ge=15, le=86400)
    background_fill_mode: str | None = None

    @field_validator("background_fill_mode")
    @classmethod
    def validate_background_fill_mode(cls, value: str | None) -> str | None:
        if value is None:
            return value
        if value not in VALID_BACKGROUND_FILL_MODES:
            raise ValueError(
                f"background_fill_mode must be one of {sorted(VALID_BACKGROUND_FILL_MODES)!r}, got {value!r}"
            )
        return value


class PlaylistItemBackgroundResponse(BaseModel):
    """Subdued background fill colours derived from the display variant."""

    ready: bool
    dominant_color: str
    gradient_colors: list[str]


class PlaylistItemResponse(BaseModel):
    asset_id: str
    filename: str
    display_url: str
    thumbnail_url: str
    width: int
    height: int
    checksum_sha256: str
    mime_type: str
    background: PlaylistItemBackgroundResponse | None = None

    @classmethod
    def from_domain(cls, item: PlaylistItem) -> "PlaylistItemResponse":
        bg: PlaylistItemBackgroundResponse | None = None
        if item.background is not None:
            bg = PlaylistItemBackgroundResponse(
                ready=item.background.ready,
                dominant_color=item.background.dominant_color,
                gradient_colors=item.background.gradient_colors,
            )
        return cls(
            asset_id=item.asset_id,
            filename=item.filename,
            display_url=item.display_url,
            thumbnail_url=item.thumbnail_url,
            width=item.width,
            height=item.height,
            checksum_sha256=item.checksum_sha256,
            mime_type=item.mime_type,
            background=bg,
        )


class DisplayPlaylistResponse(BaseModel):
    collection_id: str | None
    collection_name: str | None
    shuffle_enabled: bool
    playlist_revision: str
    background_fill_mode: str
    sleep_schedule: SleepScheduleResponse
    profile: DisplayProfileResponse
    items: list[PlaylistItemResponse]

    @classmethod
    def from_domain(cls, playlist: DisplayPlaylist) -> "DisplayPlaylistResponse":
        return cls(
            collection_id=playlist.collection_id,
            collection_name=playlist.collection_name,
            shuffle_enabled=playlist.shuffle_enabled,
            playlist_revision=playlist.playlist_revision,
            background_fill_mode=playlist.background_fill_mode,
            sleep_schedule=SleepScheduleResponse.from_domain(playlist.sleep_schedule),
            profile=DisplayProfileResponse.from_domain(playlist.profile),
            items=[PlaylistItemResponse.from_domain(item) for item in playlist.items],
        )
